### PR TITLE
KAFKA-6860: Fix NPE in Kafka Streams with EOS enabled

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 1]
```diff
@@ -169,11 +168,7 @@ public void reinitializeStateStoresForPartitions(final Collection<TopicPartition
             final int partition = getPartition(topicName);
             final TopicPartition storePartition = new TopicPartition(topicName, partition);
 
-            if (checkpointableOffsets.containsKey(storePartition)) {
-                partitionsAndOffsets.put(storePartition, checkpointableOffsets.get(storePartition));
-            } else {
-                partitionsAndOffsets.put(storePartition, -1L);
-            }
+            partitionsAndOffsets.put(storePartition, checkpointableOffsets.getOrDefault(storePartition, -1L));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:171]: This is a Java8 rewrite only.

**Comment by mjsax** [comment]
Call for review @bbejeck @vvcephei 

PR also contains some code cleanup. This should be cherry-picked to `2.0` branch.

**Comment by mjsax** [comment]
@guozhangwang I think, we should also cherry-pick to `0.11.0`, `1.0`, and `1.1` branches. WDYT?

**Review by vvcephei — APPROVED** [review]
Thanks for the patch.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java [hunk 2]
```diff
@@ -62,11 +63,14 @@ public void reinitializeStateStoresForPartitions(final Logger log,
             checkpointableOffsets.remove(topicPartition);
             storeToBeReinitialized.add(changelogTopicToStore.get(topicPartition.topic()));
         }
-        try {
-            checkpoint.write(checkpointableOffsets);
-        } catch (final IOException fatalException) {
-            log.error("Failed to write offset checkpoint file to {} while re-initializing {}: {}", checkpoint, stateStores, fatalException);
-            throw new StreamsException("Failed to reinitialize global store.", fatalException);
+
+        if (eosEnabled) {
```
Comment (bk-ko) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java]: Sorry for interrupt. I guess you might miss `!` in front of `eosEnabled`. Right ?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractStateManager.java]: Great catch! Thanks for reviewing! (this happens when stupid me does not write a unit test...)

**Comment by mjsax** [comment]
Updated this. Also added unit-test. Some more Java8 cleanup.

**Comment by bk-ko** [comment]
It would be nice if this fix applies to `1.1`!

**Comment by bbejeck** [comment]
retest this please

**Review by bbejeck — APPROVED** [review]
Thanks for the patch @mjsax, LGMT

**Review by guozhangwang — APPROVED** [review]
LGTM!

**Comment by guozhangwang** [comment]
Assuming clients:checkstyleMain is fixed from jenkins, please feel free to merge and cherry-pick to old branches.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.0` and `1.1`. (`1.0` is no affected)

**Comment by bk-ko** [comment]
Thank you for your quick fix!

**Pull request merged**